Write chapter on state-dependent and server-dependent service rates (#30)
MichalisPanayides/Thesis: chapters/06_agent_based_extension/01_state_dependent/main.tex

\section{State-dependent variation}\label{sec:state_dependent}

This section focuses on creating a state-dependent variation of the queueing
system described in Section~\ref{sec:queueing_section}.
The state-dependent version is based on the idea that the speed of the servers
could be based on the number of individuals present in the system.
This was implemented only on the Discrete Event Simulation (DES) version of

> the queueing model described in Section~\ref{sec:discrete_event_simulation}.

Reviewer: Does this need to be here? At the moment you're just describing the conceptual model right? The implementation doesn't matter.

Author: Yeah, I see. I think I mention it in the implementation section later on

> The new service rate \(\mu\) that is used here is defined as follows:

Reviewer: ```suggestion
A parametric service rate \(\mu=\mu(s)\) is defined as follows for a given \(s\in S\): % TODO Is S the right thing?
```

Author: Yep, `S` is the set of all states. Will make it `\( (u,v) \in S \)` though

\begin{equation}\label{eq:state_dependent_service_rate}
    \mu =
    \begin{cases}
        \mu_{(0,0)}, & \text{if } (u, v) = (0, 0) \\
        \mu_{(0,1)}, & \text{if } (u, v) = (0, 1) \\
        \quad \vdots & \qquad \quad \vdots \\
        \mu_{(N,M)}, & \text{if } (u, v) = (M, N)
    \end{cases}
\end{equation}

Although the state-dependent variation of the queueing system was implemented
only on the DES version of the queue, for visualisation purposes, the Markov

> chain model will be used to better demonstrate the concept of this model.

Reviewer: I think you just don't need to mention this yet. The model is conceptual at the moment.

Author: I wasn't planning on saying that, but I do use the Markov model figure somewhere and kinda felt that I should mention something that the Markov model is not actually part of the agent-based model and was a bit out of the blue.

This is the picture that I use 👇 which I think is helpful with visualisation:

![image](https://user-images.githubusercontent.com/20097997/218537037-43f0ce53-b0c9-4c99-a734-fadb23f22a47.png)

Reviewer: > I wasn't planning on saying that, but I do use the Markov model figure somewhere and kinda felt that I should mention something that the Markov model is not actually part of the agent-based model and was a bit out of the blue.

The Markov model is the basis for the abs model. It's just that you don't get the steady state probabilities analytically at any point. The conceptual model that this is a picture of is still there.

Consider an example of the queueing system described in
Section~\ref{sec:queueing_section} where the number of servers is set to
\(C = 1\), the threshold is set to \(T = 1\), node 1 capacity is \(N = 2\) and
node 2 capacity is \(M = 1\).
Figure~\ref{fig:state_dependent_markov_example} shows the state-dependent
Markov chain model of the queueing system.

\begin{figure}[H]
    \centering
    \input{chapters/06_agent_based_extension/Bin/example_model_1121/main.tex}
    \caption{Markov chain example with \(C=1, T=1, N=2, M=1\)}
    \label{fig:state_dependent_markov_example}
\end{figure}

In the previous chapters of this thesis, the value of the service rate \(\mu\)
was set to a constant value.
For this variation of the model \(\mu\) will have a different value depending
on the number of individuals present in the two nodes of the queue.
In other words \(\mu\) will take a different value depending on the state

> \(u, v\) of which the system is in.

Reviewer: Move this to before the definition of \(\mu\)

Author: Will do 👍

Consider the following value of \(\mu\) for the Markov chain model shown in
Figure~\ref{fig:state_dependent_markov_example}.

\begin{equation*}
    \mu =
    \begin{cases}
        \mu_{(0, 0)}, \text{ if } (u, v) = (0, 0) \\
        \mu_{(0, 1)}, \text{ if } (u, v) = (0, 1) \\
        \mu_{(0, 2)}, \text{ if } (u, v) = (0, 2) \\
        \mu_{(1, 1)}, \text{ if } (u, v) = (1, 1) \\
        \mu_{(1, 2)}, \text{ if } (u, v) = (1, 2)
    \end{cases}
\end{equation*}

\subsection{Implementation}

> The state-dependent variation of the queueing system was implemented using the
> python library \texttt{ciw}~\cite{ciwpython}.

Reviewer: ```suggestion
The state-dependent variation of the queueing system was implemented as a discrete event simulation model using the
python library \texttt{ciw}~\cite{ciwpython}.
```

All the code used to implement the state-dependent variation of the queueing

> system is publicly available on GitHub.

Reviewer: ```suggestion
system is archived at % TODO add archive reference
and developed openly on Github.
```

\begin{lstlisting}[style=pystyle]
>>> import ciw
>>> class StateDependentExponential(
...     ciw.dists.Distribution
... ):
...     def __init__(self, rates):
...         self.rates = rates
... 
...     def sample(self, t=None, ind=None):
...         """
...         This method is used to sample the service time for an
...         individual based on the current state
...         """
...         state = (
...             len(ind.simulation.nodes[1].individuals[0]),
...             len(ind.simulation.nodes[2].individuals[0]),
...         )
...         is_invalid_state = (
...             state[0] > 0 and state[1] < ind.simulation.threshold
...         )
...         if is_invalid_state:
...             state = (state[0] - 1, state[1] + 1)
...         rate = self.rates[state]
...         return random.expovariate(rate)
    
\end{lstlisting}

The function \texttt{simulate\_model} described in
Section~\ref{sec:discrete_event_simulation} takes \texttt{mu} as one of its
arguments.
Whenever the value of \(\mu\) is a dictionary with keys being the possible
states and values the service rate for that state, the \texttt{mu} argument
will be a \texttt{StateDependentExponential} object.

> The following code shows the implementation of the

Reviewer: Avoid `the following` use `ref` etc...

Author: Will do. I don't think I `referece` code snippets throughout the thesis. Will make a new PR to change that everywhere

Reviewer: Sorry for missing that before.
